src/transformers/models/whisper/generation_whisper.py

```diff
@@ -0,0 +1,1436 @@
+                init_tokens += [forced_decoder_ids[0][1]]
+                forced_decoder_ids = forced_decoder_ids[1:]
+                i += 1
+
+            forced_decoder_ids = forced_decoder_ids if len(forced_decoder_ids) > 0 else None
+            generation_config.forced_decoder_ids = forced_decoder_ids
+
+        return init_tokens
+
+    def _retrieve_logit_processors(
+        self, generation_config, logits_processor, no_speech_threshold, num_start_tokens, is_shortform, num_beams
+    ):
+        begin_index = 1
+        if generation_config.return_timestamps is True:
+            forced_decoder_ids = generation_config.forced_decoder_ids
+            last_forced_decoder_ids = forced_decoder_ids[-1][-1] if forced_decoder_ids is not None else None
+            if last_forced_decoder_ids == generation_config.no_timestamps_token_id:
+                # remove no_timestamp to be forcefully generated if we want to return timestamps
+                # this is also important to make sure `WhisperTimeStampLogitsProcessor` functions correctly
+                forced_decoder_ids = forced_decoder_ids[:-1] if len(forced_decoder_ids) > 1 else None
+                # Make sure that if list is empty we set it to None
+                generation_config.forced_decoder_ids = forced_decoder_ids
+
+            begin_index = begin_index + len(forced_decoder_ids) if forced_decoder_ids is not None else begin_index
+
+            timestamp_processor = WhisperTimeStampLogitsProcessor(generation_config, begin_index=begin_index)
+            logits_processor = (
+                [timestamp_processor] if logits_processor is None else [timestamp_processor] + logits_processor
+            )
+
+        if generation_config.suppress_tokens is not None:
+            suppress_tokens_processor = SuppressTokensLogitsProcessor(generation_config.suppress_tokens)
+            logits_processor = (
+                [suppress_tokens_processor]
+                if logits_processor is None
+                else [suppress_tokens_processor] + logits_processor
+            )
+            generation_config.suppress_tokens = None
+
+        if generation_config.begin_suppress_tokens is not None:
+            begin_suppress_processor = SuppressTokensAtBeginLogitsProcessor(
+                generation_config.begin_suppress_tokens, begin_index=begin_index
+            )
+            logits_processor = (
+                [begin_suppress_processor]
+                if logits_processor is None
+                else [begin_suppress_processor] + logits_processor
+            )
+            generation_config.begin_suppress_tokens = None
+
+        if no_speech_threshold is not None and not is_shortform:
+            no_speech_detector = WhisperNoSpeechDetection(
+                no_speech_token=generation_config.no_timestamps_token_id - 1,
+                begin_index=begin_index,
+                begin_index_offset=num_start_tokens,
```

Comment:
I've improved the logic here to make it easier to understand. In short what happens is the following:
- a) We have `begin_index` which defines the position of the first generated token
- b) We have `begin_index_offset` which defines the offset between the first generated token and the "begin-of-sentence-token" So for "<previous_conditioning> <sot> <en> <transcribe> <first gen token>", `begin_index_offset` would be 3 (pos diff between <sot> and <first gen token>). Note that `begin_index` is not three but much larger since <previous_conditioning> needs to be accounted for.

For the very first generation a) and b) are the same however which allows us to simplify the code here! 

Also `begin_index_offset` is a bad name, will rename it to `start_of_trans_offset`.